matrix-macros.el

```diff
@@ -36,9 +162,189 @@ Is transformed to:
 \(with-slots (id session) room
   (with-slots (user) session
     user))"
-  (declare (indent defun))
+  (declare (debug (listp body))
+           (indent defun))
   (cl-loop for (slots object) in (reverse slots-objects)
            do (setq body `((with-slots ,slots ,object ,@body)))
            finally return (car body)))
 
+(defvar-local matrix-url-with-retrieve-async-timeout-timer nil
+  "When a response buffer has a timeout, this variable stores the
+  timer object so that it may be canceled if the request
+  completes successfully.")
+
+(cl-defun matrix-url-with-retrieve-async (url &key cbargs silent inhibit-cookies data
+                                              (method "GET") extra-headers query timeout success error
+                                              parser (query-on-exit t))
+  "Retrieve URL asynchronously with `url-retrieve'.
+
+Arguments CBARGS, SILENT, and INHIBIT-COOKIES are passed to
+`url-retrieve', which see.
+
+DATA is bound to `url-request-data', which see.
+
+METHOD may be a symbol or string, which is bound as a capitalized
+string to `url-request-method', which see.
+
+EXTRA-HEADERS is an alist of header-value pairs, which is bound
+to `url-request-extra-headers', which see.
+
+QUERY is an alist of key-value pairs which is appended to the URL
+as the query.
+
+SUCCESS may be a function symbol or a body form, which is called
+with zero arguments upon successful completion of the request.
+In the call to SUCCESS, these variables will be bound:
+
+`status': See `url-retrieve'.
+`cbargs': See `url-retrieve'.
+`headers': The HTTP response headers as a string.
+`body': The HTTP response body as a string.
+
+ERROR may be a function symbol or a body form, which is called
+with zero arguments if the request fails.  In the error call,
+these variables will be bound, in addition to the ones bound for
+SUCCESS:
+
+`errors': The list of `url' error symbols for the most recent
+error, e.g. `(error http 404)' for an HTTP 404 error.
+
+In the SUCCESS and ERROR calls, the current buffer is the
+response buffer, and it is automatically killed when the call
+completes.
+
+PARSE-BODY-FN may be a function which parses the body and returns
+a value to bind `body' to.  The point is positioned after the
+headers, at the beginning of the body, before calling the
+function.  For example, `json-read' may be used to parse JSON
+documents, after which the parsed JSON would be available in
+SUCCESS and ERROR as `body'.  Or, if the body is not needed,
+`ignore' could be used to prevent the body from being parsed."
+  (declare (indent defun))
+  (let* ((success-body-fn (cl-typecase success
+                            (function success)
+                            (otherwise (byte-compile
```

Comment:
I think it is needed here, because `url-with-retrieve-async` is a function, not a macro, and it creates those lambdas at runtime, so byte-compiling the file won't compile the lambdas.